src/structures/paging/page_table.rs

```diff
@@ -89,11 +87,37 @@ impl PageTableEntry {
         assert!(!flags.contains(PageTableFlags::HUGE_PAGE));
         self.set_addr(frame.start_address(), flags)
     }
+}
 
-    /// Sets the flags of this entry.
+#[cfg(feature = "dynamic_flags")]
+impl PageTableEntry {
+    /// Returns the physical address mapped by this entry, might be zero.
     #[inline]
-    pub fn set_flags(&mut self, flags: PageTableFlags) {
-        self.entry = self.addr().as_u64() | flags.bits();
+    pub fn addr(&self) -> PhysAddr {
+        PhysAddr::new(self.entry & PHYSICAL_ADDRESS_MASK.load(Ordering::Relaxed))
+    }
+
+    /// Returns the flags of this entry.
+    #[inline]
+    pub fn flags(&self) -> PageTableFlags {
+        PageTableFlags::from_bits_retain(
+            self.entry & !PHYSICAL_ADDRESS_MASK.load(Ordering::Relaxed),
+        )
+    }
+}
+
+#[cfg(not(feature = "dynamic_flags"))]
+impl PageTableEntry {
+    /// Returns the physical address mapped by this entry, might be zero.
+    #[inline]
+    pub fn addr(&self) -> PhysAddr {
+        PhysAddr::new(self.entry & 0x000f_ffff_ffff_f000u64)
+    }
+
+    /// Returns the flags of this entry.
+    #[inline]
+    pub fn flags(&self) -> PageTableFlags {
+        PageTableFlags::from_bits_truncate(self.entry)
```

Comment:
I'd prefer it if we didn't duplicate the definitions for these functions. Instead, can we define a function `physical_address_mark()` with two implementations returning `0x000f_ffff_ffff_f000u64` or `PHYSICAL_ADDRESS_MASK.load(Ordering::Relaxed)` respectively?